File: utils/currency_conversion.py
import requests

def get_exchange_rate(source_currency, target_currency):
    url = f"https://api.exchangerate-api.com/v4/latest/{source_currency}"
    response = requests.get(url)
    
    if response.status_code != 200:
        raise Exception("Error fetching data from the API")
    
    data = response.json()
    return data['rates'].get(target_currency)

def convert_currency(source_currency, target_currency, amount):
    # Get the exchange rate
    exchange_rate = get_exchange_rate(source_currency, target_currency)
    
    if exchange_rate is None:
        raise ValueError("Invalid currency type provided.")
    
    # Calculate the converted amount
    converted_amount = amount * exchange_rate
    return converted_amount

File: splitwise-web/src/services/currency_conversion.js
import axios from "axios";

/**
 * Fetches the exchange rate between two currencies.
 * @param {string} sourceCurrency - The source currency (e.g., "USD").
 * @param {string} targetCurrency - The target currency (e.g., "EUR").
 * @returns {Promise<number>} - The exchange rate.
 */
const getExchangeRate = async (sourceCurrency, targetCurrency) => {
  const url = `https://api.exchangerate-api.com/v4/latest/${sourceCurrency}`;
  
  try {
    const response = await axios.get(url);
    const rates = response.data.rates;

    if (!rates[targetCurrency]) {
      throw new Error("Invalid currency type provided.");
    }

    return rates[targetCurrency];
  } catch (error) {
    throw new Error("Error fetching data from the API");
  }
};

/**
 * Converts an amount from one currency to another.
 * @param {string} sourceCurrency - The source currency (e.g., "USD").
 * @param {string} targetCurrency - The target currency (e.g., "EUR").
 * @param {number} amount - The amount to convert.
 * @returns {Promise<number>} - The converted amount.
 */
const convertCurrency = async (sourceCurrency, targetCurrency, amount) => {
  const exchangeRate = await getExchangeRate(sourceCurrency, targetCurrency);
  return amount * exchangeRate;
};

export { getExchangeRate, convertCurrency };
